### DISCUSSION TITLE
Align catch with switch syntax

### DISCUSSION THREAD

**Discussion by Andrew-Hanlon** [dotnet/csharplang#2809] [Language Ideas]
The existing catch block is already *patterny* with type handling and `when` conditions. 

It would be great if the catch block had alignment with the switch expression and statement syntaxes.

Example of a few proposed syntaxes:

````csharp
try { ... }
catch(Exception ex)
{                    
    case OperationCanceledException c when c.CancellationToken != token :
        throw;

    case OperationCanceledException _ :
    case InvalidOperationException  _ when state == knownState :
        return;
                    
    default:
        Log(ex);
        throw;
}


try { ... }
catch
{
    OperationCanceledException c when c.CancellationToken != token => throw;
    OperationCanceledException _                                   => return;
    InvalidOperationException  _ when state == knownState          => return;
    _                                                              => throw;
}

Or

try { ... }
catch (OperationCanceledException c) when (c.CancellationToken != token) => throw;
catch (InvalidOperationException _)  when (state == knownState)          => return;
catch (OperationCanceledException _)                                     => return;
catch (Exception _)                                                      => throw;
````

This would be an addition, not a change - and I believe would not have any issue with backward compatibility or interoperability with the existing syntax.

**Importantly** in both syntaxes the block *contents* (each case) would be equivalent to writing a set of `catch(ExType ex) when (condition) {...}` blocks. Thereby, if the default case is omitted, then standard catch fall-through is followed.  

While there is zero criticality to this proposal, it would nicely unify the language syntax - and IMHO looks sleek.

(One half of this proposal was raised already in #2688)

**Comment by Joe4evr** [comment]
I don't see how this proposal concerning writing exception handlers is in any way related to 2568 which is about XML docs.

**Comment by Andrew-Hanlon** [comment]
@Joe4evr Sorry typo in issue number. Corrected.
